cleverpush/src/main/java/com/cleverpush/banner/AppBannerModule.java

```diff
@@ -150,246 +154,253 @@ public void onFailure(int statusCode, String response, Throwable throwable) {
+
+            }
+
+            @Override
+            public void onFailure(int statusCode, String response, Throwable throwable) {
+                Log.e(TAG, "App Banner Event failed." +
+                        "\nStatus code: " + statusCode +
+                        "\nResponse: " + response
+                );
+            }
+        });
+    }
+
+    public void initSession(String channel) {
+        this.channel = channel;
+
+        if (
+                !CleverPush.getInstance(activity).isDevelopmentModeEnabled()
+                        && lastSessionTimestamp > 0
+                        && (System.currentTimeMillis() - lastSessionTimestamp) < MIN_SESSION_LENGTH
+        ) {
+            return;
+        }
+
+        if (popups.size() > 0) {
+            for (AppBannerPopup popup : popups) {
+                popup.dismiss();
+            }
+            popups = new ArrayList<>();
+        }
+
+        lastSessionTimestamp = System.currentTimeMillis();
+
+        sessions += 1;
+        this.saveSessions();
+
+        banners = null;
+        handler.post(this::loadBanners);
+        this.startup();
+    }
+
+    private int getSessions() {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(CleverPush.context);
+        return sharedPreferences.getInt(CleverPushPreferences.APP_BANNER_SESSIONS, 0);
+    }
+
+    private void saveSessions() {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(CleverPush.context);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.putInt(CleverPushPreferences.APP_BANNER_SESSIONS, sessions);
+        editor.apply();
+    }
+
+    public void triggerEvent(String key, String value) {
+        events.put(key, value);
+
+        this.startup();
+    }
+
+    public void getBanners(AppBannersListener listener) {
+        this.getBanners(listener, null);
+    }
+
+    public void getBanners(AppBannersListener listener, String notificationId) {
+        if (listener != null) {
+            if (notificationId != null) {
+                // reload banners because the banner might have been created just seconds ago
+                getBannersListeners.add(listener);
+                handler.post(() -> {
+                    this.loadBanners(notificationId);
+                });
+            } else {
+                if (banners == null) {
+                    getBannersListeners.add(listener);
+                } else {
+                    listener.ready(banners);
+                }
+            }
+        }
+    }
 
     private void startup() {
-    	Log.d(TAG, "startup");
+        Log.d(TAG, "startup");
 
-    	this.getBanners(banners -> {
-			createBanners(banners);
-			scheduleBanners();
-		});
+        this.getBanners(banners -> {
+            createBanners(banners);
+            scheduleBanners();
+        });
     }
 
     private void createBanners(Collection<Banner> banners) {
```

Comment:
please rework this whilst having different objectives in the mind:
a) KISS - Keep It Simple Stupid. Try not to over-engineer this by overusing booleans. (https://en.wikipedia.org/wiki/KISS_principle)

b) Keep the methods short. Approx not longer than 20 lines.

Reply:
We'll rework in SDK 2.0